### ISSUE TITLE
Support for G700s?

### ISSUE THREAD

**Issue by floe** [libratbag/libratbag#22]
Hi everyone, awesome work! I have a Logitech G700s mouse which currently seems to be unsupported. What would be required to add support for that device type?

**Comment by bentiss** [comment]
Hey Florian,

The G700s should be fairly similar to the G700, G500s and G500. To do the first attempts to communicate with the mouse, it should be a matter of just adding your VID/PID in driver-hidpp10.c, and see how it behaves.
I have a G700 that I could access in (roughly) one month, and then I could also have a deeper look at the hid++ 1.0 implementation we have.

**Comment by cvuchener** [comment]
I expect it to work like the G500s with maybe a different format for the resolution. You can test tools from this repo too: https://github.com/cvuchener/g500 Results from [Reading the memory](https://github.com/cvuchener/g500#reading-the-memory) and [Testing some reading request](https://github.com/cvuchener/g500#testing-some-reading-request) would be interesting.

**Comment by floe** [comment]
Did what @bentiss suggested; here's the output (which seems slightly off, although 13 buttons seems to be correct):
```
$ sudo ./ratbag-command info /dev/input/event8 
Device 'Logitech G700s Rechargeable Gaming Mouse'
Capabilities: res profile btn-key btn-macros
Number of buttons: 13
Profiles supported: 3
  Profile 0
    Resolutions:
      0: 46256x17284dpi @ 7Hz
      1: 18084x35968dpi @ 7Hz
      2: 32100x9842dpi @ 7Hz
      3: 13600x25650dpi @ 7Hz
      4: 17288x62136dpi @ 7Hz
    Button: 0 type left is mapped to 'none'
    Button: 1 type middle is mapped to 'none'
    Button: 2 type right is mapped to 'none'
    Button: 3 type thumb is mapped to 'none'
    Button: 4 type thumb2 is mapped to 'none'
    Button: 5 type thumb3 is mapped to 'none'
    Button: 6 type wheel left is mapped to 'none'
    Button: 7 type wheel right is mapped to 'none'
    Button: 8 type resolution up is mapped to 'none'
    Button: 9 type resolution down is mapped to 'none'
    Button: 10 type unknown is mapped to 'none'
    Button: 11 type unknown is mapped to 'none'
    Button: 12 type unknown is mapped to 'none'
  Profile 1
    Resolutions:
      0: 17984x17284dpi @ 7Hz
      1: <disabled>
      2: <disabled>
      3: 13600x12850dpi @ 7Hz
      4: 34576x62136dpi @ 7Hz
    Button: 0 type left is mapped to 'none'
    Button: 1 type middle is mapped to 'none'
    Button: 2 type right is mapped to 'none'
    Button: 3 type thumb is mapped to 'none'
    Button: 4 type thumb2 is mapped to 'none'
    Button: 5 type thumb3 is mapped to 'none'
    Button: 6 type wheel left is mapped to 'none'
    Button: 7 type wheel right is mapped to 'none'
    Button: 8 type resolution up is mapped to 'none'
    Button: 9 type resolution down is mapped to 'none'
    Button: 10 type unknown is mapped to 'none'
    Button: 11 type unknown is mapped to 'none'
    Button: 12 type unknown is mapped to 'none'
  Profile 2 (active)
    Resolutions:
      0: <disabled>
      1: <disabled>
      2: <disabled>
      3: 13600x36864dpi @ 7Hz
      4: 14788x12984dpi @ 7Hz
    Button: 0 type left is mapped to 'none'
    Button: 1 type middle is mapped to 'none'
    Button: 2 type right is mapped to 'none'
    Button: 3 type thumb is mapped to 'none'
    Button: 4 type thumb2 is mapped to 'none'
    Button: 5 type thumb3 is mapped to 'none'
    Button: 6 type wheel left is mapped to 'none'
    Button: 7 type wheel right is mapped to 'none'
    Button: 8 type resolution up is mapped to 'none'
    Button: 9 type resolution down is mapped to 'none'
    Button: 10 type unknown is mapped to 'none'
    Button: 11 type unknown is mapped to 'none'
    Button: 12 type unknown is mapped to 'none'
```

@cvuchener I also tested your g500 tool; I was able read a 27-page memory dump. Any suggestions where you'd like to have them uploaded?

**Comment by cvuchener** [comment]
Upload them any where you want and post the link here or on my github, or send them by email at clement dot vuchener at gmail dot com or both. 27 pages is a lot more than I got with mine, I expect the memory layout will be different.

The 13 button count is hard coded because the G500s has 13 buttons in its profile though only 10 are physically present. It does not mean it worked.

**Comment by floe** [comment]
OK, see here: http://floe.butterbrot.org/external/g700s_dump.zip

**Comment by cvuchener** [comment]
Again it is a slightly different structure. I had a look at page 2. What I
read is:
5 resolution modes:
0808 2011 resolution = 0x08, leds = off off on
1818 2012 resolution = 0x18, leds = off on on
2828 2022 resolution = 0x28, leds = on on on
4040 1022 resolution = 0x40, leds = on on off
8282 1021 resolution = 0x82, leds = on off off
Some other settings (I expect the default resolution mode and the refresh
rate to be among them)
0180 0110 0201 2c02 5864 ffbc 0009 31
13 buttons
810100 button 1
810200 button 2
810400 button 3
810800 button 4
811000 button 5
812000 button 6
814000 button 7
830300 special
830800 special dpi-
830400 special dpi+
831100 special
830100 special pan left
830200 special pan right

Does that look like one of your profiles? I guess the unknown special
functions are "switch profile" and "battery level".


2015-09-20 9:52 GMT+02:00 Florian Echtler <notifications@github.com>:

**Comment by cvuchener** [comment]
I started writing the g700s profile documentation :
https://github.com/cvuchener/g500/blob/master/doc/profile/g700s.md

About the memory layout, I found valid profiles in pages 2 to 6 and page 25
and 26. Page 4 has some macros after the profile.

2015-09-20 10:27 GMT+02:00 Clément VUCHENER <clement.vuchener@gmail.com>:

**Comment by cvuchener** [comment]
2015-09-20 10:51 GMT+02:00 Clément VUCHENER <clement.vuchener@gmail.com>:

> I started writing the g700s profile documentation :
> https://github.com/cvuchener/g500/blob/master/doc/profile/g700s.md
>
> About the memory layout, I found valid profiles in pages 2 to 6 and page
> 25 and 26. Page 4 has some macros after the profile.
>
I forgot: page 6 has macros too, and page 1 has the same kind of unknown
values found on the other mice.

**Comment by cvuchener** [comment]
I added a tool: logitech-raw-command, it should be easier to use that send-raw-request.

I also added a script using it. Could you send the output of `./scan-read-command.sh /dev/hidrawN`?

**Comment by floe** [comment]
```
$ ./scan-read-command.sh /dev/hidraw0
Error: 1
$ ./scan-read-command.sh /dev/hidraw1
Command 0x00: short
00 00 00
Command 0x01: short
00 00 00
Command 0x07: short
00 22 00
Command 0x0f: short
00 02 00
Command 0x51: short
00 00 00
Command 0x61: short
00 10 00
Command 0x63: long
14 00 14 00 80 01 01 00 00 00 00 00 00 00 00 00
Command 0x64: short
08 00 00
Command 0xa2: long
7b ff bd bf 3d bb 68 76 b2 9e a7 ff ef ef 6d ec
Command 0xd0: short
00 00 00
Command 0xd3: long - unavailable
Command 0xd5: long - unavailable
Error: 3
```

**Comment by cvuchener** [comment]
> P.S. actually, I've never used that mouse with Windows or any Logitech tool, so I don't even know how the profiles look.

You can just press them and see what happens. The buttons are numbered and we may assume there are in the same order in the profile. From what I see on the web G8 should show the battery level (830300) and G11 switch profile (831100). G9 and G10 should change the resolution mode.

Does the LED change when you switch resolution or profile? Does either `./logitech-raw-command /dev/hidraw1 0x07 read short` or `./logitech-raw-command /dev/hidraw1 0x51 read short` change with the LEDs?

When you switch profile does the output of `./logitech-raw-command /dev/hidraw1 0x0f read short` change?

We need to find how resolution is encoded. Look at `./logitech-raw-command /dev/hidraw1 0x63 read long` with different resolution. If you don't know the resolution, measure it with mouse-dpi-tool in libedev-utils or [this program](http://pastebin.com/2r2xq8SL) (trace a line of a certain length on a sheet of paper, start the program with the event dev and line length in inch, right click at the beginning of the line and left click at the end).

**Comment by cvuchener** [comment]
I wrote in this file: https://gist.github.com/cvuchener/a5a8083a1e1ffa6cd435 what I understand of the profiles in the memory dump (number between parentheses are offset (for referencing macros) not data from the memory). Do they match with what your mouse do? Also for each profile try `./logitech-raw-command /dev/hidraw1 0x64 read short`, do the output look like the "refresh rate?" value in the profiles?

**Comment by phomes** [comment]
I think the g700s is pretty interesting to look closer at. It has some feature that none of the mice I have access to have:

1) LEDs can set both color and brightness. Perhaps that is what the value in byte 3 of the G9 is for.

2) Each profile can have a shift (sniper mode) DPI level. Holding the shift button will temporarily go the the selected DPI level. Should be a value like the Default DPI setting.

3) A toggle for per-axis DPI (does the g500 have this setting?)

4) A toggle for acceleration

5) A setting for "Power mode"

If we can understand those we might be able to fill in some blanks in the docs for the g500.

**Comment by cvuchener** [comment]
> 1) LEDs can set both color and brightness. Perhaps that is what the value in byte 3 of the G9 is for.

I don't think there is a color setting: http://support.logitech.com/article/41587. And there is no color in the first three bytes of the profile. The 0x57 command is not present.

> 2) Each profile can have a shift (sniper mode) DPI level. Holding the shift button will temporarily go the the selected DPI level. Should be a value like the Default DPI setting.

Done in the driver not the hardware.

> However, the DPI Shift function is only available in Automatic Game Detection mode.

From the same page as above.

> 3) A toggle for per-axis DPI (does the g500 have this setting?)

It not a toggle in the profile structure, the resolution is simply repeated twice.

> 4) A toggle for acceleration

Maybe byte 1 in command 0x61 and byte 36 in G500 profile structure.



The unknown special button may be "DPI cycle".

**Comment by cvuchener** [comment]
The third byte in the items of page 1 match the LEDs for the profile;
```
02 00 01 -> binary 0001
03 00 03 -> binary 0011
04 00 07 -> binary 0111
05 00 06 -> binary 0110
06 00 04 -> binary 0100
ff ...
```

**Comment by cvuchener** [comment]
For testing notification bits:
- try setting all bits with `./logitech-raw-command /dev/hidraw1 0x00 write short FF FF FF`
- read what bits were effectively set `./logitech-raw-command /dev/hidraw1 0x00 read short`

For testing feature bits:
- try setting all bits with `./logitech-raw-command /dev/hidraw1 0x01 write short FF FF FF`
- read what bits were effectively set `./logitech-raw-command /dev/hidraw1 0x01 read short`

**Comment by bentiss** [comment]
@floe I just pushed the initial support of the G700 (not 's') to libratbag. Adding the G700s should be a matter of adding it to the hwdb now. Reading the profiles should work. The macros are not parsed nor the consumer control keys, but this is just an implementation detail :)

There is still no support for writing settings. I made it work on the G502, and I hope that the hid++ 1.0 devices should be also easy as soon as the writing of the profile is coded.

Thanks everyone for helping reverse engineering the memory mapping of the profiles :)
